README
k9an-wsprd is a decoder for K1JT's Weak Signal Propagation Reporter (WSPR) mode.

The program is written in C and is a command-line program that reads from a
.c2 file or .wav file and writes output to the console. It is designed so that
it can be used as a drop-in replacement for the wsprd program that wspr-x uses 
for decoding. 

k9an-wsprd requires only one external library: libfftw3

COMPILING:
A bare-bones Makefile is included. 

USAGE:
k9an-wsprd [options] infile

       -a <path> path to writeable data files, default="."
       -c write .c2 file at the end of the first pass
       -C maximum number of decoder cycles per bit
       -d deeper search. Slower, a few more decodes
       -e x (x is transceiver dial frequency error in Hz)
       -f x (x is transceiver dial frequency in MHz)
       -F use the Fano decoder instead of stack decoder
       -H do not use (or update) the hash table
       -m decode wspr-15 .wav file
       -q quick mode - doesn't dig deep for weak signals
       -s single pass mode, no subtraction (same as original wsprd)
       -v verbose mode (shows dupes)
       -w wideband mode - decode signals within +/- 150 Hz of center
       -z x (x is fano metric table bias, default is 0.42)

infile can be either .wav or .c2

e.g. 
./k9an-wsprd -wf 14.0956 140709_2258.wav

Note that for .c2 files, the frequency within the file overrides the command
line value.

FEATURES:
By default, k9an-wsprd reports signals that are within +/- 110 Hz of the
subband center frequency. The wideband option (-w) extends this to +/- 150 Hz.

k9an-wsprd maintains a hashtable and will decode all three types of wspr
messages. An option (-H) is available to turn off use of the hashtable.

The symbols are decoded using Phil Karn's sequential decoder routine,
fano.c.

NOTES:
This program attempts to maximize the number of successful decodes per transmit
interval by trying to decode virtually every peak in the averaged spectrum. 
The program also implements two-pass decoding, whereby signals that are successfully
decoded are subtracted one-by-one during the first decoding pass. Then, the 
decoder is run again. In many cases the subtraction process will uncover signals
that can then be successfully decoded on the second pass.

There will be occasional duplicate decodes when two closely spaced 
peaks come from the same signal. The program removes dupes based on callsign 
and frequency. Two decodes that have the same callsign and estimated frequencies
that are within 1 Hz will be treated as decodes of the same signal. This
dupechecking is turned off with the -v flag.

Feedback and suggestions for improvement are welcome!

Steve Franke, k9an
Urbana, IL, USA
s.j.franke@icloud.com




ChangeLog
7-20-2915

Improvements to efficiency of sync search. Added new decoder (Jelinek stack-
bucket decoder) which is more efficient than Fano decoder. Modified code to
improve the effectiveness of -O3 -ffast-math optimization.

6-22-2015

Many changes and bug fixes to k9an-wsprd. Also added new two-pass decoding 
mode which can increase the number of spots on a crowded band by as much 
as 30%. 

Also added new wsprsim program which encodes a message that is supplied
on the command line and prints out the channel symbols.

5-24-2015 

Moved all callsign unpacking into a function. Added some sanity checks
on decoded power and grid.

5-23-2015

Merged in K1JT's changes, including the new -a option to move output files.
Speed up sync_and_demodulation function by calculating sin/cos tables only
when necessary.
Improved metric table calculation, and added -z option to allow variable
metric table bias. This gives us a single "knob" which can be tuned to 
trade off probability of timeout (erasures) vs probability of bit errors.
-z 0.5 corresponds to the "standard" bias, equal to the code rate. Smaller
bias values result in higher path metrics,  making it more likely that 
the Fano algorithm will make it to the final node within the allowed 
number of cycles, but also increasing the probability that the final 
result will contain an error...

5-11-2015 Steve K9AN

The decoder now works with wspr-15 files. It will automatically detect the
mode for a .c2 file. The -m option (no argument) must be used to decode
a wspr-15 .wav file.

Also added "slow" mode (-s option). Currently, it is 20000 cycles and ii=1

5-8-2015 Joe Taylor, K1JT and Steve Franke, K9AN

Merged in Joe Taylor's modifications which yield a huge speed improvement
and also increase # of decodes by 2-3%. Speed improvement ranges from a
factor of 4 to a factor of 8 depending on noise type, with the larger 
improvements associated with files that contain significant lightning events.

3-7-2015 Steve K9AN

Fixed a problem that affected extended callsigns with a 2-character suffix.

2-20-2015 Steve K9AN

Fixed a bug that caused slight frequency error (less than 0.5 Hz). Thanks
to Glenn, N6GN for pointing me toward this issue.

Also added one more digit of precision to frequency reports in ALL_WSPR.TXT.
This will be used to study the variability of frequency estimates.

2-16-2015 Steve K9AN

Applied grid6 patch contributed by John, KD6EKQ. This fixes a bug that
caused the displayed 6-digit locator to be corrupted when there were two
or more type 3 messages decoded.

2-15-2015 Steve K9AN

Applied patch file contributed by John, KD6EKQ. This adds the following:

* Added the ability to read and incorporate fftw wisdom to speed up the
DFT transforms when using PCM WAV input files. The code checks for the
"WSPRD_FFTW_WISDOM" environment variable. If set, it specifies the
file with fftw wisdom.

* Relaxed the error checks on PCM WAV input files to allow 24, 32 bit and
float sample formats (libsndfile handles the details of the conversion,
so there is no real reason to exclude such files).

* Added the -e command line option. It enables code that corrects for
transceiver tuning error. The argument is the dial freq error in Hz.


mettab.c
/*
 This file is part of k9an-wsprd.
 
 File name: mettab.c
 Description:
 
 Copyright 2014-2015, Steven Franke, K9AN
 License: GNU GPL v3
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

int mettab[2][256]={
    5,   5,   5,   5,   5,   5,   5,   5,   5,   5,
    5,   5,   5,   5,   5,   5,   5,   5,   5,   5,
    5,   5,   5,   5,   5,   5,   5,   5,   5,   5,
    5,   5,   5,   5,   5,   5,   5,   5,   5,   5,
    5,   5,   5,   5,   5,   5,   5,   5,   5,   5,
    5,   5,   5,   5,   5,   5,   5,   5,   5,   5,
    5,   5,   5,   5,   5,   5,   5,   5,   5,   5,
    5,   5,   5,   5,   5,   5,   5,   5,   5,   4,
    4,   4,   4,   4,   4,   4,   4,   4,   4,   4,
    4,   4,   4,   4,   4,   4,   4,   4,   4,   4,
    3,   3,   3,   3,   3,   3,   3,   3,   3,   2,
    2,   2,   2,   2,   1,   1,   1,   1,   0,   0,
    -1,  -1,  -1,  -2,  -2,  -3,  -4,  -4,  -5,  -6,
    -7,  -7,  -8,  -9, -10, -11, -12, -12, -13, -14,
    -15, -16, -17, -17, -18, -19, -20, -21, -22, -22,
    -23, -24, -25, -26, -26, -27, -28, -29, -30, -30,
    -31, -32, -33, -33, -34, -35, -36, -36, -37, -38,
    -38, -39, -40, -41, -41, -42, -43, -43, -44, -45,
    -45, -46, -47, -47, -48, -49, -49, -50, -51, -51,
    -52, -53, -53, -54, -54, -55, -56, -56, -57, -57,
    -58, -59, -59, -60, -60, -61, -62, -62, -62, -63,
    -64, -64, -65, -65, -66, -67, -67, -67, -68, -69,
    -69, -70, -70, -71, -72, -72, -72, -72, -73, -74,
    -75, -75, -75, -77, -76, -76, -78, -78, -80, -81,
    -80, -79, -83, -82, -81, -82, -82, -83, -84, -84,
    -84, -87, -86, -87, -88, -89, -89, -89, -88, -87,
    -86, -87, -84, -84, -84, -83, -82, -82, -81, -82,
    -83, -79, -80, -81, -80, -78, -78, -76, -76, -77,
    -75, -75, -75, -74, -73, -72, -72, -72, -72, -71,
    -70, -70, -69, -69, -68, -67, -67, -67, -66, -65,
    -65, -64, -64, -63, -62, -62, -62, -61, -60, -60,
    -59, -59, -58, -57, -57, -56, -56, -55, -54, -54,
    -53, -53, -52, -51, -51, -50, -49, -49, -48, -47,
    -47, -46, -45, -45, -44, -43, -43, -42, -41, -41,
    -40, -39, -38, -38, -37, -36, -36, -35, -34, -33,
    -33, -32, -31, -30, -30, -29, -28, -27, -26, -26,
    -25, -24, -23, -22, -22, -21, -20, -19, -18, -17,
    -17, -16, -15, -14, -13, -12, -12, -11, -10,  -9,
    -8,  -7,  -7,  -6,  -5,  -4,  -4,  -3,  -2,  -2,
    -1,  -1,  -1,   0,   0,   1,   1,   1,   1,   2,
    2,   2,   2,   2,   3,   3,   3,   3,   3,   3,
    3,   3,   3,   4,   4,   4,   4,   4,   4,   4,
    4,   4,   4,   4,   4,   4,   4,   4,   4,   4,
    4,   4,   4,   4,   5,   5,   5,   5,   5,   5,
    5,   5,   5,   5,   5,   5,   5,   5,   5,   5,
    5,   5,   5,   5,   5,   5,   5,   5,   5,   5,
    5,   5,   5,   5,   5,   5,   5,   5,   5,   5,
    5,   5,   5,   5,   5,   5,   5,   5,   5,   5,
    5,   5,   5,   5,   5,   5,   5,   5,   5,   5,
    5,   5,   5,   5,   5,   5,   5,   5,   5,   5,
    5,   5,   5,   5,   5,   5,   5,   5,   5,   5,
    5,   5};
